cranelift/codegen/src/ir/dfg.rs

```diff
@@ -45,6 +45,41 @@ impl IndexMut<Inst> for Insts {
     }
 }
 
+/// Storage for basic blocks within the DFG.
+#[derive(Clone, PartialEq, Hash)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub struct Blocks(PrimaryMap<Block, BlockData>);
+
+impl Blocks {
+    /// Create a new basic block.
+    pub fn make_block(&mut self) -> Block {
+        self.0.push(BlockData::new())
+    }
+
+    /// Get the total number of basic blocks created in this function, whether they are
+    /// currently inserted in the layout or not.
+    ///
+    /// This is intended for use with `SecondaryMap::with_capacity`.
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns `true` if the given block reference is valid.
+    pub fn is_valid(&self, block: Block) -> bool {
+        self.0.is_valid(block)
+    }
+
+    /// Get the number of parameters on `block`.
+    pub fn num_block_params(&self, block: Block, pool: &ValueListPool) -> usize {
```

Comment:
Others may disagree with me here but I don't think there's value in the `num_block_params` interface. Getting the slice of block params doesn't cost any more than getting its length (I think it adds 4 to a pointer it already computed to get the length, and returns the adjusted pointer along with the length) so I don't think the extra API clutter is worth its weight.

We can leave `DataFlowGraph::num_block_params` so we aren't forcing people to change their calls for no reason, but I'd rather not add this function in this new spot.

Reply:
That sounds good to me!